File: clients/go/encryption.go
package zama

import (
	"context"
	"errors"
	"fmt"
	"math/big"

	"github.com/ethereum/go-ethereum/common"
	pb "github.com/zama-ai/sdk/clients/go/v3/internal/gen/zama/sdk/v1beta1"
	"google.golang.org/grpc"
)

type EncryptInput struct {
	wire *pb.EncryptInput
	err  error
}

type EncryptParams struct {
	Values          []EncryptInput
	ContractAddress common.Address
	UserAddress     common.Address
}

type EncryptOptions struct {
	// TimeoutMS is the SDK relayer timeout, not the ctx deadline; nil keeps the SDK default.
	TimeoutMS *uint32
}

type EncryptResult struct {
	EncryptedValues []common.Hash
	InputProof      []byte
}

func Ebool(value bool) EncryptInput {
	return EncryptInput{wire: &pb.EncryptInput{Value: &pb.EncryptInput_Ebool{Ebool: value}}}
}

func EboolBigInt(value *big.Int) EncryptInput {
	if value == nil {
		return EncryptInput{err: errors.New("ebool value is nil")}
	}
	return EncryptInput{wire: &pb.EncryptInput{Value: &pb.EncryptInput_EboolBigint{EboolBigint: value.String()}}}
}

func Euint8(value *big.Int) EncryptInput {
	if value == nil {
		return EncryptInput{err: errors.New("euint8 value is nil")}
	}
	return EncryptInput{wire: &pb.EncryptInput{Value: &pb.EncryptInput_Euint8{Euint8: value.String()}}}
}

func Euint16(value *big.Int) EncryptInput {
	if value == nil {
		return EncryptInput{err: errors.New("euint16 value is nil")}
	}
	return EncryptInput{wire: &pb.EncryptInput{Value: &pb.EncryptInput_Euint16{Euint16: value.String()}}}
}

func Euint32(value *big.Int) EncryptInput {
	if value == nil {
		return EncryptInput{err: errors.New("euint32 value is nil")}
	}
	return EncryptInput{wire: &pb.EncryptInput{Value: &pb.EncryptInput_Euint32{Euint32: value.String()}}}
}

func Euint64(value *big.Int) EncryptInput {
	if value == nil {
		return EncryptInput{err: errors.New("euint64 value is nil")}
	}
	return EncryptInput{wire: &pb.EncryptInput{Value: &pb.EncryptInput_Euint64{Euint64: value.String()}}}
}

func Euint128(value *big.Int) EncryptInput {
	if value == nil {
		return EncryptInput{err: errors.New("euint128 value is nil")}
	}
	return EncryptInput{wire: &pb.EncryptInput{Value: &pb.EncryptInput_Euint128{Euint128: value.String()}}}
}

func Euint256(value *big.Int) EncryptInput {
	if value == nil {
		return EncryptInput{err: errors.New("euint256 value is nil")}
	}
	return EncryptInput{wire: &pb.EncryptInput{Value: &pb.EncryptInput_Euint256{Euint256: value.String()}}}
}

func Eaddress(value common.Address) EncryptInput {
	return EncryptInput{wire: &pb.EncryptInput{Value: &pb.EncryptInput_Eaddress{Eaddress: value.Bytes()}}}
}

func (s *SDKContext) Encrypt(ctx context.Context, params EncryptParams, options EncryptOptions) (*EncryptResult, error) {
	values := make([]*pb.EncryptInput, len(params.Values))
	for i, value := range params.Values {
		if value.err != nil {
			return nil, fmt.Errorf("encryption input %d: %w", i, value.err)
		}
		if value.wire == nil {
			return nil, fmt.Errorf("encryption input %d: missing value", i)
		}
		values[i] = value.wire
	}
	response, err := call(ctx, s, func(ctx context.Context, op *pb.Operation, trailer grpc.CallOption) (*pb.EncryptResponse, error) {
		return s.client.rpc.Encrypt(ctx, &pb.EncryptRequest{
			Operation:       op,
			Values:          values,
			ContractAddress: params.ContractAddress.Bytes(),
			UserAddress:     params.UserAddress.Bytes(),
			TimeoutMs:       options.TimeoutMS,
		}, trailer)
	})
	if err != nil {
		return nil, err
	}
	if len(response.EncryptedValues) != len(params.Values) {
		return nil, errors.New("encrypted value count does not match inputs")
	}
	result := &EncryptResult{EncryptedValues: make([]common.Hash, len(response.EncryptedValues)), InputProof: response.InputProof}
	for i, value := range response.EncryptedValues {
		hash, err := hashFromWire(value, "invalid encrypted value in response")
		if err != nil {
			return nil, err
		}
		result.EncryptedValues[i] = hash
	}
	return result, nil
}

File: clients/rust/src/encryption.rs
use crate::{Address, B256, BigInt, Sdk, generated, types::handle};
use crate::{ClientError, Result};

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum EncryptInput {
    Bool(bool),
    BoolBigInt(BigInt),
    Uint8(BigInt),
    Uint16(BigInt),
    Uint32(BigInt),
    Uint64(BigInt),
    Uint128(BigInt),
    Uint256(BigInt),
    Address(Address),
}

#[derive(Clone, Copy, Debug)]
pub struct EncryptParams<'a> {
    pub values: &'a [EncryptInput],
    pub contract_address: Address,
    pub user_address: Address,
}

#[derive(Clone, Copy, Debug, Default)]
pub struct EncryptOptions {
    /// Relayer timeout, not the RPC deadline; None keeps the SDK default.
    pub timeout_ms: Option<u32>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct EncryptResult {
    pub encrypted_values: Vec<B256>,
    pub input_proof: Vec<u8>,
}

impl From<&EncryptInput> for generated::EncryptInput {
    fn from(input: &EncryptInput) -> Self {
        use generated::encrypt_input::Value;
        let value = match input {
            EncryptInput::Bool(value) => Value::Ebool(*value),
            EncryptInput::BoolBigInt(value) => Value::EboolBigint(value.to_string()),
            EncryptInput::Uint8(value) => Value::Euint8(value.to_string()),
            EncryptInput::Uint16(value) => Value::Euint16(value.to_string()),
            EncryptInput::Uint32(value) => Value::Euint32(value.to_string()),
            EncryptInput::Uint64(value) => Value::Euint64(value.to_string()),
            EncryptInput::Uint128(value) => Value::Euint128(value.to_string()),
            EncryptInput::Uint256(value) => Value::Euint256(value.to_string()),
            EncryptInput::Address(value) => Value::Eaddress(value.to_vec()),
        };
        Self { value: Some(value) }
    }
}

impl Sdk {
    pub async fn encrypt(
        &self,
        params: EncryptParams<'_>,
        options: EncryptOptions,
    ) -> Result<EncryptResult> {
        let response = rpc!(
            self,
            encrypt,
            EncryptRequest {
                values: params.values.iter().map(Into::into).collect(),
                contract_address: params.contract_address.to_vec(),
                user_address: params.user_address.to_vec(),
                timeout_ms: options.timeout_ms,
            }
        )
        .await?;
        if response.encrypted_values.len() != params.values.len() {
            return Err(ClientError::protocol(
                "encrypted value count does not match inputs",
            ));
        }
        Ok(EncryptResult {
            encrypted_values: response
                .encrypted_values
                .iter()
                .map(|value| handle(value))
                .collect::<Result<_>>()?,
            input_proof: response.input_proof,
        })
    }
}
